px-oci-mon/main.go

```
@@ -0,0 +1,259 @@
+		pxImage = ociInstallerImage
+	}
+
+	err := validateMounts(mntFileName, dockerFileSockName)
+	if err != nil {
+		usage(err)
+	}
+
+	id, err := utils.GetMyContainerID()
+	if err != nil {
+		logrus.WithError(err).Error("Could not determine my container ID" +
+			" - are you running me inside Docker?")
+		os.Exit(-1)
+	}
+
+	di, err := utils.NewDockerInstaller(os.Getenv("REGISTRY_USER"), os.Getenv("REGISTRY_PASS"))
+	if err != nil {
+		logrus.WithError(err).Error("Could not 'talk' to Docker")
+		usage("Could not 'talk' to Docker" +
+				" - please restart using '-v /var/run/docker.sock:/var/run/docker.sock' option")
+	}
+
+	opts, err := di.ExtractConfig(id)
+	if err != nil {
+		logrus.WithError(err).Error("Could not extract my container's configuration" +
+			" - are you running me inside Docker?")
+		os.Exit(-1)
+	}
+
+	// TODO: Sanity checks for options
+	logrus.Debugf("OPTIONS:: %+v\n", opts)
+
+	err = installPxFromOciImage(di, pxImage, opts)
+	if err != nil {
+		logrus.WithError(err).Error("Could not install Portworx service")
+		os.Exit(-1)
+	}
+
+	logrus.Info("Stopping Portworx service (if any)")
+	err = runExternal("/bin/sh", "-c", "systemctl stop portworx")
+	logrus.WithError(err).Debugf("Stopping done")
+
+	logrus.Info("Starting Portworx service")
+	err = runExternal("/bin/sh", "-c",
+		`systemctl daemon-reload && systemctl enable portworx && systemctl start portworx`)
+	if err != nil {
+		logrus.WithError(err).Error("Could not start Portworx service")
+		os.Exit(-1)
+	}
+}
+
+func doUninstall() {
+	logrus.Info("Stopping Portworx service")
+	err := runExternal("/bin/sh", "-c",
+		`systemctl stop portworx`)
+	if err != nil {
+		logrus.WithError(err).Error("Could not stop Portworx service")
+		os.Exit(-1) // NOTE: CRITICAL failure !!
+	}
+
+	logrus.Info("Disabling Portworx service")
+	err = runExternal("/bin/sh", "-c",
+		`systemctl disable portworx`)
+	if err != nil {
+		logrus.WithError(err).Warn("Could not disable Portworx service (continuing)")
```

Looks like the approach here is best-effort.

if we failed to disable the service, let's not move forward. A service that isn't disabled will attempt to spin up on the next systemd start. If we ignore the error here, it will give the end user a false positive that portworx is uninstalled from the system, while in reality, the service is still enabled.

---

ok.

---

Actually -- the service files, and the binaries will be removed, so the service won't be able to start, but let's not risk weird service errors -- will change this to critical failure.